core/src/duplicate_shred.rs

```diff
@@ -0,0 +1,338 @@
+    slot: Slot,
+    shred_index: u32,
+    shred_type: ShredType,
+    // Serialized DuplicateSlotProof split into chunks.
+    num_chunks: u8,
+    chunk_index: u8,
+    #[serde(with = "serde_bytes")]
+    chunk: Vec<u8>,
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct DuplicateShredIndex {
+    slot: Slot,
+    shred_index: u32,
+    shred_type: ShredType,
+    num_chunks: u8,
+    chunk_index: u8,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("data chunk mismatch")]
+    DataChunkMismatch,
+    #[error("decoding error")]
+    DecodingError(std::io::Error),
+    #[error("encoding error")]
+    EncodingError(std::io::Error),
+    #[error("invalid chunk index")]
+    InvalidChunkIndex,
+    #[error("invalid duplicate shreds")]
+    InvalidDuplicateShreds,
+    #[error("invalid duplicate slot proof")]
+    InvalidDuplicateSlotProof,
+    #[error("invalid size limit")]
+    InvalidSizeLimit,
+    #[error("invalid shred")]
+    InvalidShred(#[from] ShredError),
+    #[error("number of chunks mismatch")]
+    NumChunksMismatch,
+    #[error("missing data chunk")]
+    MissingDataChunk,
+    #[error("(de)serialization error")]
+    SerializationError(#[from] bincode::Error),
+    #[error("shred index mismatch")]
+    ShredIndexMismatch,
+    #[error("shred type mismatch")]
+    ShredTypeMismatch,
+    #[error("slot mismatch")]
+    SlotMismatch,
+    #[error("value out of bounds")]
+    ValueOutOfBounds,
+}
+
+// Verifies that the two shreds can indicate duplicate proof.
+fn check_shreds(shred1: &Shred, shred2: &Shred) -> Result<(), Error> {
+    if shred1.slot() != shred2.slot() {
+        Err(Error::SlotMismatch)
+    } else if shred1.index() != shred2.index() {
+        Err(Error::ShredIndexMismatch)
+    } else if shred1.common_header.shred_type != shred2.common_header.shred_type {
+        Err(Error::ShredTypeMismatch)
+    } else if shred1.payload == shred2.payload {
+        Err(Error::InvalidDuplicateShreds)
+    } else {
+        Ok(())
+    }
+}
+
+/// Splits a DuplicateSlotProof into DuplicateShred
+/// chunks with a size limit on each chunk.
+pub fn from_duplicate_slot_proof<F>(
+    proof: &DuplicateSlotProof,
+    self_pubkey: Pubkey,
+    wallclock: u64,
+    max_size: usize, // Maximum serialized size of each DuplicateShred.
+    encoder: F,
+) -> Result<impl Iterator<Item = DuplicateShred>, Error>
+where
+    F: FnOnce(Vec<u8>) -> Result<Vec<u8>, std::io::Error>,
+{
+    if proof.shred1 == proof.shred2 {
+        return Err(Error::InvalidDuplicateSlotProof);
+    }
+    let shred1 = Shred::new_from_serialized_shred(proof.shred1.clone())?;
+    let shred2 = Shred::new_from_serialized_shred(proof.shred2.clone())?;
+    check_shreds(&shred1, &shred2)?;
+    let (slot, shred_index, shred_type) = (
+        shred1.slot(),
+        shred1.index(),
+        shred1.common_header.shred_type,
+    );
+    let data = bincode::serialize(proof)?;
+    let data = encoder(data).map_err(Error::EncodingError)?;
```

Comment:
generally, I think we should avoid `map_err` as it squashes information inside the inner error. Can we wrap the `std::io::Error` and bubble it up here?

Reply:
ah I missed that it was already wrapped, that works!